{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0",
   "metadata": {},
   "source": [
    "This notebook builds Extended Data Fig. 6, MAF vs. Fitness Score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import altair as alt\n",
    "import numpy as np\n",
    "from scipy.stats import mannwhitneyu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "file = './Data/final_tables/supplementary_file_1_BARD1_SGE_final_table.xlsx' #SGE datafile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_data(file):\n",
    "    df = pd.read_excel(file, sheet_name = 'scores')\n",
    "    df = df.loc[~df['variant_qc_flag'].isin(['WARN'])]\n",
    "\n",
    "    raw_data = df\n",
    "    \n",
    "    gnomad_df = df.dropna(subset = ['gnomad_af'])\n",
    "    gnomad_df['dataset'] = 'gnomAD'\n",
    "    gnomad_df = gnomad_df.drop(columns = ['regeneron_maf'])\n",
    "    \n",
    "    reg_df = df.dropna(subset = ['regeneron_maf'])\n",
    "    reg_df = reg_df.drop(columns = ['gnomad_af'])\n",
    "    reg_df['dataset'] = 'Regeneron'\n",
    "    \n",
    "    gnomad_df = gnomad_df.rename(columns = {'gnomad_af': 'Allele Frequency'})\n",
    "    reg_df = reg_df.rename(columns = {'regeneron_maf': 'Allele Frequency'})\n",
    "\n",
    "\n",
    "    df = pd.concat([gnomad_df, reg_df])\n",
    "\n",
    "    df = df.rename(columns = {'consequence': 'Consequence', 'score': 'snv_score_minmax'}) #Renames columns to harmonize column names\n",
    "\n",
    "    #Renames variant consequences to be cleaner\n",
    "    df.loc[df['Consequence'].str.contains('missense'), 'Consequence'] = 'Missense'\n",
    "    df.loc[df['Consequence'] == 'synonymous_variant', 'Consequence'] = 'Synonymous'\n",
    "    df.loc[df['Consequence'] == 'intron_variant', 'Consequence'] = 'Intron'\n",
    "    df.loc[df['Consequence'] == 'stop_gained', 'Consequence'] = 'Stop Gained'\n",
    "    df.loc[df['Consequence'] == 'stop_lost', 'Consequence'] = 'Stop Lost'\n",
    "    df.loc[df['Consequence'].str.contains('site'), 'Consequence'] = 'Canonical Splice'\n",
    "    df.loc[df['Consequence'].str.contains('ing_var'), 'Consequence'] = 'Splice Region'\n",
    "    df.loc[df['Consequence'].str.contains('UTR'), 'Consequence'] = 'UTR Variant'\n",
    "    df.loc[df['Consequence'] == 'start_lost', 'Consequence'] = 'Start Lost'\n",
    "\n",
    "    df['log_AF'] = np.log10(df['Allele Frequency'])\n",
    "\n",
    "    return df, raw_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def heatmap(df):\n",
    "\n",
    "    \n",
    "    plot = alt.Chart(df).mark_rect().encode(\n",
    "        x = alt.X('snv_score_minmax:Q',\n",
    "                  axis = alt.Axis(\n",
    "                      title = 'Fitness Score',\n",
    "                      titleFontSize = 20,\n",
    "                      labelFontSize = 18,\n",
    "                      values = [-0.4, -0.2, 0]\n",
    "                  ),\n",
    "                  bin = alt.Bin(maxbins = 50)\n",
    "                 ),\n",
    "        y = alt.Y('log_AF:Q',\n",
    "                  bin = alt.Bin(maxbins = 20),\n",
    "                  axis = alt.Axis(\n",
    "                      title = 'log10(Allele Frequency)',\n",
    "                      titleFontSize = 20,\n",
    "                      labelFontSize = 18,\n",
    "                      values = [0, -1, -2, -3, -4, -5, -6]\n",
    "                  )\n",
    "                  ),\n",
    "        color = alt.Color('count():Q',\n",
    "                          scale = alt.Scale(scheme = 'lighttealblue',\n",
    "                                           domain = [1,400]),\n",
    "                          legend = alt.Legend(title = '# of Variants',\n",
    "                                              titleFontSize = 16,\n",
    "                                              labelFontSize = 14\n",
    "                                             )\n",
    "                         )\n",
    "    ).properties(\n",
    "        width = 300,\n",
    "        height = 250,\n",
    "        title = alt.TitleParams( text = 'Allele Frequency vs. Fitness Score' + ' (n = ' + str(len(df)) + ')', fontSize = 22)\n",
    "    ).configure_axis(\n",
    "        grid = False\n",
    "    ).configure_view(\n",
    "        stroke = None\n",
    "    )\n",
    "\n",
    "    plot.display()\n",
    "\n",
    "    return plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def whitneytesting(df):\n",
    "\n",
    "    gnomad_df = df.dropna(subset = ['gnomad_af']).copy()\n",
    "    reg_df = df.dropna(subset = ['regeneron_maf']).copy()\n",
    "\n",
    "    \n",
    "    gnomad_ab_maf = gnomad_df.loc[gnomad_df['functional_consequence'].isin(['functionally_abnormal'])]['gnomad_af']\n",
    "    gnomad_normal_maf = gnomad_df.loc[gnomad_df['functional_consequence'].isin(['functionally_normal'])]['gnomad_af']\n",
    "\n",
    "    reg_ab_maf = reg_df.loc[reg_df['functional_consequence'].isin(['functionally_abnormal'])]['regeneron_maf']\n",
    "    reg_normal_maf = reg_df.loc[reg_df['functional_consequence'].isin(['functionally_normal'])]['regeneron_maf']\n",
    "\n",
    "    pairs = [(gnomad_ab_maf, gnomad_normal_maf, 'gnomAD'), (reg_ab_maf, reg_normal_maf, 'Regeneron')]\n",
    "\n",
    "\n",
    "    for pair in pairs:\n",
    "        ab, normal, dataset = pair\n",
    "        stat, p_value = mannwhitneyu(ab, normal, alternative = 'less')\n",
    "\n",
    "        print(f' Testing MAF distribution for LoF vs. functionally normal variants in {dataset} data. p-value is: {p_value}. The median MAF for LoF vars. is:{ab.median()} and median for normal vars. is: {normal.median()}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main(save = True):\n",
    "    merged, testing_input = read_data(file)\n",
    "    binned_map = heatmap(merged)\n",
    "    whitneytesting(testing_input)\n",
    "\n",
    "    if save:\n",
    "        binned_map.save('/Users/ivan/Desktop/BARD1_draft_figs/fig_4a_MAF.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "main(save = False) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
